clang/lib/CIR/Dialect/Transforms/TargetLowering/RecordLayoutBuilder.cpp

```diff
@@ -10,3 +10,640 @@
+    HasPackedField = true;
+  }
+}
+
+//===-----------------------------------------------------------------------==//
+// Misc. Helper Functions
+//===----------------------------------------------------------------------===//
+
+bool isMsLayout(const CIRLowerContext &Context) {
+  return Context.getTargetInfo().getCXXABI().isMicrosoft();
+}
+
+/// Does the target C++ ABI require us to skip over the tail-padding
+/// of the given class (considering it as a base class) when allocating
+/// objects?
+static bool mustSkipTailPadding(clang::TargetCXXABI ABI, const StructType RD) {
+  assert(!::cir::MissingFeatures::recordDeclIsCXXDecl());
+  switch (ABI.getTailPaddingUseRules()) {
+  case clang::TargetCXXABI::AlwaysUseTailPadding:
+    return false;
+
+  case clang::TargetCXXABI::UseTailPaddingUnlessPOD03:
+    // FIXME: To the extent that this is meant to cover the Itanium ABI
+    // rules, we should implement the restrictions about over-sized
+    // bitfields:
+    //
+    // http://itanium-cxx-abi.github.io/cxx-abi/abi.html#POD :
+    //   In general, a type is considered a POD for the purposes of
+    //   layout if it is a POD type (in the sense of ISO C++
+    //   [basic.types]). However, a POD-struct or POD-union (in the
+    //   sense of ISO C++ [class]) with a bitfield member whose
+    //   declared width is wider than the declared type of the
+    //   bitfield is not a POD for the purpose of layout.  Similarly,
+    //   an array type is not a POD for the purpose of layout if the
+    //   element type of the array is not a POD for the purpose of
+    //   layout.
+    //
+    //   Where references to the ISO C++ are made in this paragraph,
+    //   the Technical Corrigendum 1 version of the standard is
+    //   intended.
+    // FIXME(cir): This always returns true since we can't check if a CIR record
+    // is a POD type.
+    assert(!::cir::MissingFeatures::CXXRecordDeclIsPOD());
+    return true;
+
+  case clang::TargetCXXABI::UseTailPaddingUnlessPOD11:
+    // This is equivalent to RD->getTypeForDecl().isCXX11PODType(),
+    // but with a lot of abstraction penalty stripped off.  This does
+    // assume that these properties are set correctly even in C++98
+    // mode; fortunately, that is true because we want to assign
+    // consistently semantics to the type-traits intrinsics (or at
+    // least as many of them as possible).
+    llvm_unreachable("NYI");
+  }
+
+  llvm_unreachable("bad tail-padding use kind");
+}
+
+} // namespace
+
+/// Get or compute information about the layout of the specified record
+/// (struct/union/class), which indicates its size and field position
+/// information.
+const CIRRecordLayout &CIRLowerContext::getCIRRecordLayout(const Type D) const {
+  assert(isa<StructType>(D) && "Not a record type");
+  auto RT = dyn_cast<StructType>(D);
+
+  assert(RT.isComplete() && "Cannot get layout of forward declarations!");
+
+  // FIXME(cir): Cache the layout. Also, use a more MLIR-based approach.
```

Comment:
"use a more MLIR-based approach", in respect to what? not clear how actionable this is, please be more clarify by being more specific